### PR TITLE
LPS-X Test to reveal the Whip bug

### PR DESCRIPTION
**Pull request by dantewang**
@shuyangzhou This test will fail if compiled and executed on JDK 1.8.0_161+. Not sure if modules/test is the right place to put the test.

### CODE REVIEW

**Comment by shuyangzhou** [comment]
@dantewang the newer version jdk changed how finally block's bytecode is generated.

It is basically duplicating the finally body to 2 pieces. A is dispatched when the try block does not throw exception, B is dispatched when the try block does throw an exception.

B has some additional bytecode, at beginning does a local var store to keep the reference of the exception object, at end does a load to fetch back the exception for throwing.

The missing line from report is the exception rethrowing at the end of the finally block.

Considering the real logic of the finally body is duplicated, for the sake of code coverage purpose, as long as the try body throws an exception, we will have a complete coverage. (Not necessary enough for the business logic testing purpose, just for coverage.)

So a work around in your test case is simply:
```
public class TryFinallyClass {

	public String tryFinally() {
		try {
			return toString();
		}
		finally {
			hashCode();
		}
	}

	@Override
	public String toString() {
		throw new IllegalStateException();
	}

}
```
and 
```
	@Test
	public void testTryFinally() {
		TryFinallyClass tryFinallyClass = new TryFinallyClass();

		try {
			tryFinallyClass.tryFinally();

			Assert.fail();
		}
		catch (IllegalStateException ise) {
		}
	}
```
So basically we need to find a way to make the try block throwing an exception. If it is too hard to do in real case, then we should really reconsider why there is a try block here.

Anyway, this is not something we can or should improve from whip. Either we make try block fail from test, or we remove the try block from production code.

Do we still have unit tests having issue with this?

**Comment by dantewang** [comment]
@shuyangzhou If my memory is not wrong, this issue was found when trying to cover the first try-finally block in `com.liferay.portal.cache.internal.dao.orm.EntityCacheImpl#loadResult`. The second try-finally can be covered just fine. But as far as I can I remember, I tried throwing exception from `openSession()` but the block was still not fully covered. Is it also related to whether there is a `return` in the try block?

**Comment by shuyangzhou** [comment]
Are we still trying to update the tests to cover that? If so we can put it into the context, then it will be easier. I am closing this pull now, send me a follow up if needed.
